tools/bridge/tests/test_transfer_fetcher.py

```diff
@@ -1,23 +1,235 @@
+    def transfer(receiver):
+        token_contract.functions.transfer(receiver, 1).transact(
+            {"from": premint_token_address}
         )
-        assert queue.get() == 0
-        assert queue.empty()
-        gevent.sleep(0.3)
-        assert queue.empty()
-        tester_foreign.mine_block()
-        gevent.sleep(0.3)
-        assert queue.get() == 1
-        assert queue.empty()
-    finally:
-        transfer_event_fetcher.kill()
+
+    return transfer
+
+
+@pytest.fixture
+def transfer_tokens_to_foreign_bridge(foreign_bridge_contract, transfer_tokens_to):
+    def transfer():
+        transfer_tokens_to(foreign_bridge_contract.address)
+
+    return transfer
+
+
+def test_instantiate_transfer_fetcher_with_non_existing_token_contract(
+    token_transfer_event_queue,
+    w3_foreign,
+    foreign_bridge_contract,
+    max_reorg_depth,
+    event_filter_fetch_limit,
+):
+    with pytest.raises(ValueError):
+        TransferFetcher(
+            queue=token_transfer_event_queue,
+            w3_foreign=w3_foreign,
+            token_contract_address="0x0000000000000000000000000000000000000000",
+            foreign_bridge_contract_address=foreign_bridge_contract.address,
+            max_reorg_depth=max_reorg_depth,
+            event_filter_fetch_limit=event_filter_fetch_limit,
+        )
+
+
+def test_instantiate_transfer_fetcher_with_incorrect_token_contract(
+    token_transfer_event_queue,
+    w3_foreign,
+    foreign_bridge_contract,
+    event_filter_fetch_limit,
+):
+    with pytest.raises(ValueError):
+        TransferFetcher(
+            queue=token_transfer_event_queue,
+            w3_foreign=w3_foreign,
+            token_contract_address=foreign_bridge_contract.address,
+            foreign_bridge_contract_address=foreign_bridge_contract.address,
+            max_reorg_depth=max_reorg_depth,
+            event_filter_fetch_limit=event_filter_fetch_limit,
+        )
+
+
+def test_instantiate_transfer_fetcher_with_non_existing_foreign_bridge_contract(
+    token_transfer_event_queue,
+    w3_foreign,
+    token_contract,
+    max_reorg_depth,
+    event_filter_fetch_limit,
+):
+    with pytest.raises(ValueError):
+        TransferFetcher(
+            queue=token_transfer_event_queue,
+            w3_foreign=w3_foreign,
+            token_contract_address=token_contract.address,
+            foreign_bridge_contract_address="0x0000000000000000000000000000000000000000",
+            max_reorg_depth=max_reorg_depth,
+            event_filter_fetch_limit=event_filter_fetch_limit,
+        )
+
+
+def test_fetch_token_transfer_events_in_range(
+    transfer_fetcher,
+    token_transfer_event_queue,
+    transfer_tokens_to_foreign_bridge,
+    w3_foreign,
+):
+    assert token_transfer_event_queue.empty()
+
+    transfer_tokens_to_foreign_bridge()
+    transfer_fetcher._fetch_token_transfer_events_in_range(
+        0, w3_foreign.eth.blockNumber
+    )
+
+    assert token_transfer_event_queue.qsize() == 1
```

Comment:
please also test that it's the right event that has been fetched.

Reply:
Good point
:heavy_check_mark: